IoT-Projects/ESP8266/NeoPixel Smart Planter/UBIDOTS_ESP8266_RGB_SMART_PLANTER.ino

```diff
@@ -0,0 +1,266 @@
+/*
+
+RGB Smart Planter integrated with Ubidots for Monitoring & Control. 
+
+This code works for:
+
+    1) Read two sensors: DHT11, and Soil Moisture.
+    2) Publish sensors readings to Ubidots.
+    3) Subscribe to multiple variables for remote control.
+
+Libraries required:
+
+- Ubidots ESP8266 MQTT - (https://github.com/ubidots/ubidots-mqtt-esp)
+- Adafruit NeoPixel - (https://github.com/adafruit/Adafruit_NeoPixel)
+- DHT - (https://github.com/adafruit/DHT-sensor-library)
+
+Made by: Maria Hernández - IoT Developer Advocate @ Ubidots
+Revision: José García - Development & Support Manager @ Ubidots
+
+*/
+/****************************************
+ * Include Libraries
+ ****************************************/
+#include "UbidotsESPMQTT.h"
+#include <Adafruit_NeoPixel.h>
+#include "DHT.h"
+#include <map>
+
+/****************************************
+ * Define Pins
+ ****************************************/
+#define LIGHTPIN D1 // Digital pin for Led Lamp.
+#define DHTPIN D5     // Digital pin for DHT sensor.
+#define NEOPIXELSPIN D6 // Digital pin for NeoPixel Ring.
+#define MOISTUREPIN A0 // Analog pin for Moisture Sensor.
+
+/****************************************
+ * Define Constants
+ ****************************************/
+#define TOKEN "BBFF-xxxxxxxxxxxxxxxxxxxxxxx" // Assign your Ubidots TOKEN.
+#define WIFINAME "xxxxxxxxxxxxx" // Assign your SSID.
+#define WIFIPASS "xxxxxxxxxxxxx" // Assign your WiFi Password.
+#define DEVICE "planter" // Ubidots Device Label.
+#define VAR_PUB_1 "temperature" // Ubidots Variables' label for publishing data.
+#define VAR_PUB_2 "humidity"
+#define VAR_PUB_3 "soil-moisture"
+#define VAR_PUB_4 "heat-index"
+#define VAR_SUB_1 "light-1" // Ubidots Variables' label for subscribing to data; These variables have to be created at Ubidots.
+#define VAR_SUB_2 "light-2" 
+#define NUMPIXELS 12 // 12 bit NeoPixel Ring
+// Uncomment whatever type you're using
+#define DHTTYPE DHT11   // DHT 11
+//#define DHTTYPE DHT22   // DHT 22  (AM2302), AM2321
+//#define DHTTYPE DHT21   // DHT 21 (AM2301)
+                       //   R,   G,   B
+uint8_t myColors[][6] = {{250,   0,   0},   // Red.
+                         {  0, 255,   0},   // Green.
+                         {  0,   0, 255},   // Blue.
+                         {255, 255,   0},   // Yellow.
+                         {255, 255, 255},   // White.                         
+                         {  0,   0,  0}};  // OFF.
+const uint8_t NUMBER_OF_VARIABLES = 2; // Number of variables for subscription.
+char * variable_labels[NUMBER_OF_VARIABLES] = {VAR_SUB_1, VAR_SUB_2}; // Variables' label for subscription.
+const int ERROR_VALUE = 65535;  // Set here an error value.
+float value; // To store incoming value.
+bool bottom_light; // flag to control conditions for the bottom light.
+
+// Comparison functor to map functions.
+struct cmp_str {
```

Comment:
1. Use camel case and follow Google' style https://google.github.io/styleguide/cppguide.html. Correct the remaining variables that are not declared using camel case.

2. This structure is useless, use a raw function

Reply:
"Compare" is the third parameter of the map class. Basically, a binary predicate that takes two element keys as arguments and returns a bool. 

This struct has to be defined anytime you want a map key to be a pointer, so you need to give a comparison functor to the map otherwise it's comparing the pointer, not the null-terminated string it points to.

Reply:
https://stackoverflow.com/questions/4157687/using-char-as-a-key-in-stdmap/4157753#4157753